nifi-nar-bundles/nifi-windows-event-log-bundle/nifi-windows-event-log-processors/src/main/java/org/apache/nifi/processors/windows/event/log/ConsumeWindowsEventLog.java

```diff
@@ -0,0 +1,282 @@
+            .build();
+
+    public static final List<PropertyDescriptor> PROPERTY_DESCRIPTORS = Collections.unmodifiableList(Arrays.asList(CHANNEL, QUERY, MAX_BUFFER_SIZE, MAX_EVENT_QUEUE_SIZE));
+
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("Relationship for successfully consumed events.")
+            .build();
+
+    public static final Set<Relationship> RELATIONSHIPS = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(REL_SUCCESS)));
+    public static final String APPLICATION_XML = "application/xml";
+    private final WEvtApi wEvtApi;
+    private final Kernel32 kernel32;
+    private final ErrorLookup errorLookup;
+    private final String name;
+
+    private Throwable wEvtApiError = null;
+    private Throwable kernel32Error = null;
+
+    private BlockingQueue<String> renderedXMLs;
+    private WEvtApi.EVT_SUBSCRIBE_CALLBACK evtSubscribeCallback;
+    private WinNT.HANDLE subscriptionHandle;
+    private ProcessSessionFactory sessionFactory;
+    private String provenanceUri;
+
+    /**
+     * Framework constructor
+     */
+    public ConsumeWindowsEventLog() {
+        this(null, null);
+    }
+
+    /**
+     * Constructor that allows injection of JNA interfaces
+     *
+     * @param wEvtApi  event api interface
+     * @param kernel32 kernel interface
+     */
+    public ConsumeWindowsEventLog(WEvtApi wEvtApi, Kernel32 kernel32) {
+        this.wEvtApi = wEvtApi == null ? loadWEvtApi() : wEvtApi;
+        this.kernel32 = kernel32 == null ? loadKernel32() : kernel32;
+        this.errorLookup = new ErrorLookup(this.kernel32);
+        if (kernel32 != null) {
+            name = Kernel32Util.getComputerName();
+        } else {
+            // Won't be able to use the processor anyway because native libraries didn't load
+            name = null;
+        }
+    }
+
+    private WEvtApi loadWEvtApi() {
+        try {
+            return WEvtApi.INSTANCE;
+        } catch (Throwable e) {
+            wEvtApiError = e;
+            return null;
+        }
+    }
+
+    private Kernel32 loadKernel32() {
+        try {
+            return Kernel32.INSTANCE;
+        } catch (Throwable e) {
+            kernel32Error = e;
+            return null;
+        }
+    }
+
+    /**
+     * Register subscriber via native call
+     *
+     * @param context the process context
+     */
+    @OnScheduled
+    public void onScheduled(ProcessContext context) throws Exception {
+        String channel = context.getProperty(CHANNEL).getValue();
+        String query = context.getProperty(QUERY).getValue();
+
+        renderedXMLs = new LinkedBlockingQueue<>(context.getProperty(MAX_EVENT_QUEUE_SIZE).asInteger());
+        provenanceUri = new URI("winlog", name, "/" + channel, query, null).toASCIIString();
+
+        evtSubscribeCallback = new EventSubscribeXmlRenderingCallback(getLogger(), s -> {
+            try {
+                renderedXMLs.put(s);
```

Comment:
The documentation states that while the callback is running, it will block further events, implying that it will not spawn extra threads for each callback.  [1][2]  I took this to mean that it would make a best-effort attempt not to miss events on a single thread.  We could set the EvtSubscribeStrict flag and log ERROR_EVT_QUERY_RESULT_STALE callbacks to let the user know it's happening.

I'd rather not add another place where we could possibly drop events if possible, since it seems like the api itself already has a release valve.

[1] https://msdn.microsoft.com/en-us/library/windows/desktop/aa385577(v=vs.85).aspx
[2] http://stackoverflow.com/questions/6901540/evtsubscribe-and-threading#answer-6903656